### ISSUE TITLE
NetLogo and Apache Log4j

### ISSUE THREAD

**Issue by steipatr** [NetLogo/NetLogo#2001]
Hi all,

Our university ITSEC team have blocked NetLogo on all machines due to the recent Apache Log4j vulnerability (https://www.ncsc.gov.uk/news/apache-log4j-vulnerability). They have identified that NetLogo also uses a version of `log4j-core-*.jar`. Is this a security concern, and if so, what can we do to get NetLogo back online?

**Comment by LaCuneta** [comment]
Hey Patrick,

Thanks for checking on this, it's good to have an issue for it where we can post the info.  NetLogo is not significantly impacted by the log4j vulnerability:

- NetLogo uses a version of log4j outside the affected version ranges.  From [the log4j post on the Apache site](https://logging.apache.org/log4j/2.x/security.html), "Log4j 1.x mitigation: Log4j 1.x is not impacted by this vulnerability."  NetLogo uses version 1.2.17.
- NetLogo does not run a server component by default that would expose log4j to input from remote users.  There is no way for someone to send the specially crafted message to NetLogo across the network if you just run normal NetLogo or NetLogo 3D models.  [The HubNet feature](http://ccl.northwestern.edu/netlogo/docs/hubnet.html) does have a server component, but it only runs when special HubNet models are used, so you'll know if you're doing that.  
- The [logging feature](http://ccl.northwestern.edu/netlogo/docs/logging.html) is what uses log4j, so if you don't enable that no messages will be logged.  It is not enabled by default.

I hope that helps convince your ITSEC team that NetLogo is safe.  Let me know if you have further questions.  Thanks!

**Comment by LaCuneta** [comment]
Adding one more item I thought of: in addition to HubNet being the only network server component in NetLogo, if a computer is not setup to receive messages from outside addresses, the risk is also further minimized.  For most user computers, I would not expect their network firewalls to allow ports 9173-9180 (the HubNet ports) to accept inbound traffic from the internet.  For servers available to the public internet, making sure these ports are blocked would be sufficient to mitigate the risk from an install of NetLogo.  The typical use case for installing NetLogo on a server system is to run BehaviorSpace experiments on computers with more processors and memory, so HubNet wouldn't be needed at all.

**Comment by mverouden** [comment]
Hi @LaCuneta, I am a direct colleague of @steipatr at Wageningen University and Research. Thank you for your explanation.

However, I am struggling with a few things still, before I will communicate the above with the ITSEC team at WUR. After scanning a computer with NetLogo 6.2.2 ITSEC found the following files in NetLogo 6.2.2\app\extensions\.bundled\vid :

- log4j-1.2.17.jar
- log4j-api-2.11.0.jar
- log4j-core-2.11.0.jar

First thing, that I find rather odd, is the fact that NetLogo uses Log4j  1.x (as seen in log4j-1.2.17.jar above). The Log4j site states: "Please note that Log4j 1.x has reached end of life and is no longer supported. Vulnerabilities reported after August 2015 against Log4j 1.x were not checked and will not be fixed. Users should upgrade to Log4j 2 to obtain security fixes.".  This versions has been end of life for **more than 6** years.

Second thing, which confuses me, is the presence of the version 2 files of log4j-api and log4j-core. How can it be that these are vesion 2? The log4j-core-2.11.0.jar file is the file, that is mainly causing the vulnerability and ought to be updated. Therefore, I strongly recommend to have another look at it.

I agree with your assessment on using HubNet. Since the users at WUR only use NetLogo on isolated computers, there won't any security risk for WUR.

**Comment by krfurlong** [comment]
Hi @LaCuneta!
**tldr; Is the NetLogo team planning on patching the `log4j` issue either through upgrading to a later version or by removing it as a dependency to the core NL application?**
Sorry in advance for digging up an old issue. I'm supporting some federal agencies in the US who have previously had NetLogo installed on their servers for parameter tuning/V&V purposes. Much like the scenario that @steipatr detailed, when the `log4j` vulnerability surfaced, these agencies blocked NetLogo. 
I've pointed them towards this thread's awesome explanation of how/where NL uses `log4j`, but I'm getting the impression that they're not so interested in the nuances of the problem and are instead of the mindset: *"any `log4j 1.x` or `log4j 2.x` dependency isn't allowed.*
I'm definitely the middle-man/developer here. I'm convinced the program is safe and I've never touched the logging feature. Just relaying information back and forth to try to get our models running again!
Thanks!!

**Comment by LaCuneta** [comment]
@krfurlong Yes, we do plan on getting log4j updated or replaced, hopefully before the next NetLogo release, which is why this issue is still open.  We do have limited developer resources so I don't have a timeline on when that'll be complete.

I understand about the cautious security admins.  Is there any extra lockdown they can do to the NetLogo app to alleviate their concerns, like blacklisting it for network communication?  I'll try to think of any other simple remediations for it.

**Comment by shaochieh72** [comment]
@LaCuenta 
Do you have an time line at least patch the log4j version?  I believe it is not a difficult update for a developer.  I am forcing to uninstall the application on our network at this time.  V/R

**Comment by LaCuneta** [comment]
@shaochieh72 

> Do you have an time line at least patch the log4j version? I believe it is not a difficult update for a developer. I am forcing to uninstall the application on our network at this time. V/R

@shaochieh72 There is no patch for the version 1.2 of log4j that NetLogo uses, as far as I'm aware (v1 is out of support).  If you know of one, please pass it along and I'll take a look.  Again, per [Apache's security docs](https://logging.apache.org/log4j/2.x/security.html), the v1 vulnerability is much less severe than the v2 and it's mitigated if you don't use the JNDI configuration or `JMSAppender`, which NetLogo does not by default, as [I mentioned out previously](https://github.com/NetLogo/NetLogo/issues/2001#issuecomment-994825272):

> Log4j 1.x does not have Lookups so the risk is lower. Applications using Log4j 1.x are only vulnerable to this attack when they use JNDI in their configuration. A separate CVE (CVE-2021-4104) has been filed for this vulnerability. To mitigate: Audit your logging configuration to ensure it has no JMSAppender configured. Log4j 1.x configurations without JMSAppender are not impacted by this vulnerability.

If you're referring to the v2 of log4j brought in by the Vid extension, I can take a look at remediating that, as it should be simpler than a v1 to v2 upgrade for the main NetLogo.  Again, there is minimal risk with that extension as well, especially if it's not used or loaded, but you can take care of that in the meantime as described [in this comment](https://github.com/NetLogo/NetLogo/issues/2001#issuecomment-995870425).

I'm curious why you still think there is a risk brought on by NetLogo that merits uninstalling the software given the info provided in this issue?  If there is some aspect of the risk I'm not aware I definitely would appreciate being informed.

**Comment by LaCuneta** [comment]
We've completed removing log4j from NetLogo (#2036), and the next release will not include the library at all.
